src/evm_circuit/arithmetic/add.rs

```diff
@@ -0,0 +1,353 @@
+                // carry_in = 0 always
+                let carry_out = meta.query_advice(bytes[1], rot_carry_bit);
+                let sum = meta.query_advice(bytes[0], rot_sum);
+
+                let lhs = first_summand + second_summand;
+                let rhs = sum + carry_out * F::from_u64(1 << 8);
+
+                exprs.push(lhs - rhs);
+            }
+
+            // i = 1..=30
+            // a_i + b_i + carry[i] = c_i + carry[i + 1] * 256
+            for idx in 1..=30 {
+                let first_summand = meta.query_advice(bytes[idx], rot_first_summand);
+                let second_summand = meta.query_advice(bytes[idx], rot_second_summand);
+                let carry_in = meta.query_advice(bytes[idx], rot_carry_bit);
+                let carry_out = meta.query_advice(bytes[idx + 1], rot_carry_bit);
+                let sum = meta.query_advice(bytes[idx], rot_sum);
+
+                let lhs = first_summand + second_summand + carry_in;
+                let rhs = sum + carry_out * F::from_u64(1 << 8);
+
+                exprs.push(lhs - rhs);
+            }
+
+            // i = 31  // carry_out = carry[0]
+            // a_31 + b_31 + carry[31] = c_31 + carry[0] * 256
+            {
+                let first_summand = meta.query_advice(bytes[31], rot_first_summand);
+                let second_summand = meta.query_advice(bytes[31], rot_second_summand);
+                let carry_in = meta.query_advice(bytes[31], rot_carry_bit);
+                // Since the first carry-in is always zero, we use that cell to
+                // witness the last carry-out.
+                let carry_out = meta.query_advice(bytes[0], rot_carry_bit);
+                let sum = meta.query_advice(bytes[31], rot_sum);
+
+                let lhs = first_summand + second_summand + carry_in;
+                let rhs = sum + carry_out * F::from_u64(1 << 8);
+
+                exprs.push(lhs - rhs);
+            }
+
+            exprs.into_iter().map(move |expr| q_group_4.clone() * expr)
+        });
+
+        let q_compress = meta.selector();
+        let byte_lookup = meta.fixed_column();
+        let word_config = WordConfig::configure(meta, r, q_compress, bytes, byte_lookup);
+
+        Self {
+            r,
+            q_group_4,
+            op,
+            global_counter,
+            stack_pointer,
+            word_config,
+        }
+    }
+
+    // a + b = c
+    pub(crate) fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        a: [Option<u8>; 32],
+        b: [Option<u8>; 32],
+        c: [Option<u8>; 32],
+        carry: [Option<bool>; 32],
+    ) -> Result<(Word<F>, Word<F>, Word<F>), Error> {
+        #[cfg(test)]
+        for idx in 0..31 {
+            let a = a[idx].unwrap_or(0);
+            let b = b[idx].unwrap_or(0);
+            let c = c[idx].unwrap_or(0);
+            let carry_in = carry[idx].unwrap_or(false);
+            let carry_out = carry[idx + 1].unwrap_or(false);
```

Comment:
`idx = 31` assertion is missing, would checking like this way be more expressive?

```suggestion
        for idx in 0..32 {
            let a = a[idx].unwrap_or(0);
            let b = b[idx].unwrap_or(0);
            let c = c[idx].unwrap_or(0);
            let carry_in = if idx == 0 { None } else { carry[idx] }.unwrap_or(false);
            let carry_out = if idx == 31 { carry[0] } else { carry[idx + 1] }.unwrap_or(false);
```